### PR TITLE
fix(api-tests): Test based on mock HERE traffic data

### LINKED ISSUE #1403: Here routing tests using mock data from #1393
- for now API tests are needed

### PR DESCRIPTION
**Pull request by aoles**
### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [ ] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [x] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [ ] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes https://github.com/GIScience/openrouteservice/issues/1403.
Also: some further improvements on #1430 

### Information about the changes

API test for the functionality of traffic-enabled routing based on HERE-like traffic data.

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

**Comment by aoles** [comment]
Not sure why the API test still fails on the GitHub runner, as for me it passes locally.

**Comment by aoles** [comment]
The API test fails on the GitHub runner with the following:
```
Error:    ResultTest.testTrafficSpeed:3656 1 expectation failed.
JSON path routes[0].summary.duration doesn't match.
Expected: is a numeric value within <1.0> of <495.0>
  Actual: <477.9>
```
The failure could be related to the fact that the server on which the API test is being executed uses a different time zone than the internally hard coded "Europe/Berlin".

**Comment by aoles** [comment]
~So far I had no luck in reproducing the error locally by changing the local time zone on the machine on which the test is being executed.~

With the "fix" introducing multiple snapping points it fails for me locally too 🙈

**Comment by aoles** [comment]
With the updated test dataset the reported difference is now:

```
Error:    ResultTest.testTrafficSpeed:3656 1 expectation failed.
JSON path routes[0].summary.duration doesn't match.
Expected: is a numeric value within <1.0> of <495.0>
  Actual: <337.0>
```

File: Dockerfile [hunk 1]
```diff
@@ -91,6 +91,7 @@ WORKDIR ${BASE_FOLDER}
 # Copy over the needed bits and pieces from the other stages.
 COPY --chown=ors:ors --from=build /ors-core/openrouteservice/target/ors.war ${BASE_FOLDER}/ors-core/ors.war
 COPY --chown=ors:ors --from=build /ors-core/openrouteservice/src/main/resources/ors-config.json ${BASE_FOLDER}/ors-core/ors-config.json
+COPY --chown=ors:ors --from=build /ors-core/openrouteservice/src/main/resources/log4j.properties ${BASE_FOLDER}/tomcat/lib/log4j.properties
```
Comment (MichaelsJP) [inline Dockerfile]:
```suggestion
```

Comment (MichaelsJP) [reply Dockerfile]: Seems to be a duplicate

File: openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * AbstractMapMatcher using Graphhopper's MapMatching internally.
+ */
+public class GhMapMatcher extends AbstractMapMatcher {
+    public GhMapMatcher(GraphHopper graphHopper) {
+        setGraphHopper(graphHopper);
+    }
+    @Override
+    public RouteSegmentInfo[] match(Coordinate[] locations, boolean bothDirections) {
+        PMap hints = new PMap()
+                .putObject("profile", "car_ors_fastest")
```
Comment (MichaelsJP) [inline openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java]: As discussed. Matching should happen per profile and be stored in the related graph folder. Matching with the HGV graph could be faster, due to the reduced graph size.

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java [hunk 3]
```diff
@@ -650,8 +651,8 @@ public RouteSegmentInfo[] getMatchedSegmentsInternal(Geometry geometry,
                                                          int trafficLinkFunctionalClass,
                                                          boolean bothDirections,
                                                          int matchingRadius) {
-        if (mMapMatcher == null || mMapMatcher.getClass() != HiddenMarkovMapMatcher.class) {
-            mMapMatcher = new HiddenMarkovMapMatcher();
+        if (mMapMatcher == null /*|| mMapMatcher.getClass() != HiddenMarkovMapMatcher.class*/) {
+			mMapMatcher = new GhMapMatcher(this); //HiddenMarkovMapMatcher();
```
Comment (MichaelsJP) [inline openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]:
```suggestion
        if (mMapMatcher == null) {
			mMapMatcher = new GhMapMatcher(this);
```

**Review by MichaelsJP — CHANGES_REQUESTED** [review]

File: openrouteservice/pom.xml [hunk 4]
```diff
@@ -591,6 +591,12 @@
             <artifactId>jackson-annotations</artifactId>
             <version>${fasterxml.jackson.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
+            <version>RELEASE</version>
+            <scope>compile</scope>
+        </dependency>
```
Comment (MichaelsJP) [inline openrouteservice/pom.xml]:
```suggestion
```
We need the jetbrains annotations?
Probably the @NotNull annotation.

File: openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        long time = mr.getMatchMillis();
+        Geometry geometry = null; // TODO: get geometry from match result
+                                  // it is unclear, whether the geometry is
+                                  // used at all, so we leave it for now.
+        RouteSegmentInfo[] rsi = new RouteSegmentInfo[1];
+        rsi[0] = new RouteSegmentInfo(edgeStates, distance, time, geometry);
+        return rsi;
+    }
+
+    @NotNull
```
Comment (MichaelsJP) [inline openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java]:
```suggestion
```
This is only relevant for IDEs. Please remove alongside the annotations package.

**Review by MichaelsJP — CHANGES_REQUESTED** [review]

File: openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java (hunk 5 above)
Comment (takb) [reply openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java]: done

File: openrouteservice/pom.xml (hunk 4 above)
Comment (takb) [reply openrouteservice/pom.xml]: done

File: openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java (hunk 2 above)
Comment (takb) [reply openrouteservice/src/main/java/org/heigit/ors/mapmatching/GhMapMatcher.java]: should be covered by https://github.com/GIScience/openrouteservice/pull/1449/commits/bad66977a25621b1162f5ae301a727e29751be9b

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java (hunk 3 above)
Comment (takb) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: got moved and cleaned up diuring https://github.com/GIScience/openrouteservice/pull/1449/commits/bad66977a25621b1162f5ae301a727e29751be9b

**Review by MichaelsJP — APPROVED** [review]

**Comment by MichaelsJP** [comment]
LGTM

**Pull request merged**